cpp/src/arrow/compute/kernels/temporal_internal.h

```diff
@@ -40,15 +40,68 @@ using arrow_vendored::date::year_month_day;
 using arrow_vendored::date::zoned_time;
 using std::chrono::duration_cast;
 
+// https://howardhinnant.github.io/date/tz.html#Examples
+using ArrowTimeZone = std::variant<const time_zone*, OffsetZone>;
+
+// Utility struct for creating overloaded visitor functions from multiple lambdas.
+// Inherits from all provided callable types and exposes their operator().
+// Usage: auto visitor = overloads{[](time_zone*){...}, [](OffsetZone){...}};
+// See https://en.cppreference.com/w/cpp/utility/variant/visit for reference
+template <class... Ts>
+struct overloads : Ts... {
+  using Ts::operator()...;
+};
+template <class... Ts>
+overloads(Ts...) -> overloads<Ts...>;
+
+template <class Duration, class Func>
+auto ApplyTimeZone(const ArrowTimeZone& tz, sys_time<Duration> st, Func&& func)
+    -> decltype(func(zoned_time<Duration>{})) {
+  if (tz.index() == 0) {
```

Comment:
Would it be better to use a visitor here?

Reply:
If you can then definitely.

Reply:
See https://github.com/apache/arrow/pull/12865/commits/11a0f085035625a92f2cab074106a8d651b7eee9

Reply:
Ping @pitrou